File: src/routes/chat.router.ts
import { AgentConfig } from '@/types/agent.interface';
import StateManager from '@/state/state-manager';
import SupervisorAgent from '@/agents/coordination/supervisor.agent';
import { generateId } from 'ai';
import express, { Router, Request, Response } from 'express';
import dotenv from 'dotenv';
import {
  ChatCompletionRequest,
  ChatCompletion,
  APIError,
  GenerationParams,
  Message,
} from '@/types/router.interface';
import { BaseState } from '@/types/state.interface';

// Load environment variables from .env file
dotenv.config();

const router: Router = Router();

// Mock database for conversation history
const conversations = new Map<string, Message[]>();

router.post('/test', async (req, res) => {
  res.json({ message: 'Test endpoint is working!' });
});

// Chat completions endpoint
router.post(
  '/completions',
  async (
    req: Request<{}, {}, ChatCompletionRequest>,
    res: Response<ChatCompletion | APIError>
  ): Promise<any> => {
    try {
      const {
        model,
        messages,
        temperature = 1,
        top_p = 1,
        n = 1,
        stream = false,
        max_tokens,
        presence_penalty = 0,
        frequency_penalty = 0,
        user,
      } = req.body;

      console.log(messages);

      // Validate required fields
      if (!model || !messages || !Array.isArray(messages)) {
        return res.status(400).json({
          error: {
            message: "Missing required fields: 'model' and 'messages' array",
            type: 'invalid_request_error',
            code: 'invalid_request',
          },
        });
      }

      // Validate message format
      for (const message of messages) {
        if (!message.role || !message.content) {
          return res.status(400).json({
            error: {
              message: "Each message must have 'role' and 'content' fields",
              type: 'invalid_request_error',
              code: 'invalid_request',
            },
          });
        }
      }

      console.log('Generating response...');

      // Generate a response
      const response = await generateResponse(messages, {
        temperature,
        top_p,
        max_tokens,
        presence_penalty,
        frequency_penalty,
      });

      // Handle streaming response
      if (stream) {
        res.setHeader('Content-Type', 'text/event-stream');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('Connection', 'keep-alive');

        for await (const chunk of streamResponse(messages, {
          model,
          temperature,
          top_p,
          max_tokens,
          presence_penalty,
          frequency_penalty,
        })) {
          res.write(`data: ${JSON.stringify(chunk)}\n\n`);
        }
        res.write('data: [DONE]\n\n');
        res.end();
        return;
      }

      // Create completion response
      const completion: ChatCompletion = {
        id: `chatcmpl-${generateId()}`,
        object: 'chat.completion',
        created: Math.floor(Date.now() / 1000),
        model,
        choices: [
          {
            index: 0,
            message: {
              role: 'assistant',
              content: response,
            },
            finish_reason: 'stop',
          },
        ],
        usage: {
          prompt_tokens: calculateTokens(messages),
          completion_tokens: calculateTokens(response),
          total_tokens: calculateTokens(messages) + calculateTokens(response),
        },
      };

      res.json(completion);
    } catch (error) {
      console.error('Error generating response:', error);
      res.status(500).json({
        error: {
          message: 'Internal server error',
          type: 'server_error',
          code: 'internal_error',
        },
      });
    }
  }
);

function calculateTokens(text: string | Message[]): number {
  // Implement your token counting logic here
  // This is a simple approximation
  if (Array.isArray(text)) {
    return text.reduce(
      (acc, msg) => acc + Math.ceil(msg.content.length / 4),
      0
    );
  }
  return Math.ceil(text.length / 4);
}

async function generateResponse(
  messages: Message[],
  params: GenerationParams
): Promise<string> {
  interface ChatState extends BaseState {
    status: string;
  }

  const stateManager = new StateManager<ChatState>({
    status: 'idle',
    messages: messages.map((msg) => {
      const role = (() => {
        switch (msg.role) {
          case 'system':
            return 'system';
          case 'user':
            return 'user';
          case 'function':
          case 'assistant':
          default:
            return 'assistant';
        }
      })();
      return { role, content: msg.content };
    }),
    internalMessages: [],
  });

  const config: AgentConfig = {
    name: 'api-supervisor',
    description: 'api supervisor',
    defaultModel: 'gpt-4o',
  };

  const agent = new SupervisorAgent(config, stateManager);

  const response = await agent.execute(messages[0].content);

  console.log(response);

  return response;
}

interface StreamChunk {
  id: string;
  object: 'chat.completion.chunk';
  created: number;
  model: string;
  choices: {
    index: number;
    delta: Partial<Message>;
    finish_reason:
      | 'stop'
      | 'length'
      | 'function_call'
      | 'content_filter'
      | null;
  }[];
}

async function* streamResponse(
  messages: Message[],
  params: GenerationParams & { model: string }
): AsyncGenerator<StreamChunk> {
  const response = await generateResponse(messages, params);
  const chunks = response.split(' ');

  for (const chunk of chunks) {
    yield {
      id: `chatcmpl-${generateId()}`,
      object: 'chat.completion.chunk',
      created: Math.floor(Date.now() / 1000),
      model: params.model,
      choices: [
        {
          index: 0,
          delta: {
            content: chunk + ' ',
          },
          finish_reason: null,
        },
      ],
    };
  }
}

export default router;

File: aurite-mcp/src/routes/chat_router.py
from __future__ import annotations

import asyncio
import json
import logging
import time
import uuid

from fastapi import APIRouter
from pydantic import BaseModel
from starlette.responses import StreamingResponse

from src.clients.client import MCPClient
from src.clients.client_handler import get_client, clear_all_clients

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/chat", tags=["chat"])


# data models
class Message(BaseModel):
    role: str
    content: str


class ChatCompletionRequest(BaseModel):
    model: str | None = None
    messages: list[Message]
    max_tokens: int | None = 512
    temperature: float | None = 0.1
    stream: bool | None = False


async def _resp_async_generator(text_resp: str, model):
    tokens = text_resp.split(" ")

    # Generate a unique ID for this completion
    completion_id = f"chatcmpl-{uuid.uuid4()}"

    # Initial chunk with role
    chunk = {
        "id": completion_id,
        "object": "chat.completion.chunk",
        "created": int(time.time()),
        "model": model,
        "system_fingerprint": "fp_" + uuid.uuid4().hex[:12],
        "choices": [{"index": 0, "delta": {"role": "assistant", "content": ""}, "logprobs": None, "finish_reason": None}],
    }
    yield f"data: {json.dumps(chunk)}\n\n"

    # Content chunks
    for token in tokens:
        chunk = {
            "id": completion_id,
            "object": "chat.completion.chunk",
            "created": int(time.time()),
            "model": model,
            "system_fingerprint": "fp_" + uuid.uuid4().hex[:12],
            "choices": [{"index": 0, "delta": {"content": token + " "}, "logprobs": None, "finish_reason": None}],
        }
        yield f"data: {json.dumps(chunk)}\n\n"
        await asyncio.sleep(0.01)

    # Final chunk
    chunk = {
        "id": completion_id,
        "object": "chat.completion.chunk",
        "created": int(time.time()),
        "model": model,
        "system_fingerprint": "fp_" + uuid.uuid4().hex[:12],
        "choices": [{"index": 0, "delta": {}, "logprobs": None, "finish_reason": "stop"}],
    }
    yield f"data: {json.dumps(chunk)}\n\n"

    # Stream termination
    yield "data: [DONE]\n\n"


@router.post("/completions")
async def chat_completions(request: ChatCompletionRequest):
    #logger.info(f"request recieved: {request}")
    
    client = await get_client(request.model)
    logger.info(client)
    logger.info(request.messages[-1].content)

    if client:
        try:
            resp_content = await client.process_query(request.messages[-1].content)
            await clear_all_clients()
        except Exception as e:
            resp_content = f"Error occured: {str(e)}"
    else:
        resp_content = "Model not recognized"

    if request.stream:
        return StreamingResponse(_resp_async_generator(resp_content, request.model), media_type="text/event-stream")

    return {
        "id": "0",
        "object": "chat.completion",
        "created": time.time(),
        "model": request.model,
        "choices": [{"message": Message(role="assistant", content=resp_content)}],
    }
